### DISCUSSION TITLE
What is the procedure for configuring DDNS service in Tasmota?

### DISCUSSION THREAD

**Discussion by rtek1000** [arendst/Tasmota#14317] [Q&A] [answered]
It appears that there is no reference to DDNS in the documentation.

I understand the concern about not using a cloud service, but having Internet access can be very useful,

DDNS can be a good alternative to cloud service.

I saw that there are Android apps for controlling Tasmota, but they only work on the local network? These apps don't support DDNS?

I found these instructions for setting up a DVR, I believe it's similar to setting up any IoT device:

> 1. Establish an account with a third-party DDNS provider.
> 
> 2. Select a username and password.
> 
> * To Setup DDNS on a DVR that is directly connected to the Internet:
> 
> 3. Go to the DVR System Setup -> Network menu.
> 
> 4. Enable DDNS. Then, enter all the DDNS info, including:
> a. Username (issued by the DDNS provider)
> b. Password (needed for the DVR to report the IP to the DDNS provider)
> c. Domain (the DDNS provider portion of the name server address)
> 
> 5. Once you save the changes, the DVR will restart, and report its current IP to your DDNS provider. The name server address should be functional within 2 minutes.
> 
> * To Setup DDNS on a DVR that is on a LAN behind a router:
> 
> 6. Make sure that your router supports DDNS.
> 
> 7. Make sure that Port Forwarding has been set up on the router so that the incoming traffic on Ports 80 and 6100 are directed to the Private (LAN) IP Address of the DVR.
> 
> 8. Go into the Web-based Setup Program of your router, and find the DDNS Setup section.
> 
> 9. Enter the DDNS information, similarly to what is outlined in Step 4.
> 
> 10. Once you save the changes, the router should restart, and report its current IP to
> your DDNS provider. The name server address should be functional within 2 minutes.

https://ganzsecurity.com/subpage/962/what-is-the-procedure-to-setup-ddns-service-with-my-digimaster-dvr

> How does it work?
> When you visit any web site on the Internet, you can either type in the domain name or you can enter the IP address, like www.geekboots.com. There are DNS servers all over the world that keep track of which names correspond to which IP addresses.
> 
> When you visit a website, your computer first looks up the IP address from a DNS server and then requests the site. This is best for servers that have static IP addresses, but not well suited for servers that are on home broadband connections.
> 
> The solution to setup dynamic DNS, whereby a third party server, which can be paid or free to keeps track of what its current IP address is and gives out this information when you type in your unique domain name. In this way, you can give your dynamic IP address a virtual static IP address. When connecting to your computer, you would either type in the domain name or you would ping the domain name to find the current IP address.
> 
> In order to determine your current IP address, all of these services work in one of two ways: either you install a piece of software on your computer that constantly updates their servers or you configure the service on your router. A lot of newer routers have DynDNS built-in, so you can have the router automatically update the IP address to the dynamic DNS provider when it changes.

https://www.geekboots.com/story/how-does-no-ip-works

> ESP8266 & ESP32 DDNS Update Client Library.
> A simple library that implements a DDNS Update Client for ESP8266 & ESP32. It is designed to be universal and currently supports DuckDNS, No-ip, DynDNS, Dynu.

https://www.arduino.cc/reference/en/libraries/easyddns/

Although the discussion is about how to access the Internet, it is interesting that the user reads these recommendations, it is a very long text, so I will only quote the introduction:

> Whenever you add devices to your network you generate additional points of potential intrusion. This is not only valid for your mobile phones and computers, but also for you Smart TV, you Alexa, or all of your SONOFF devices (ESP8266).

https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/

**Comment by Jason2866** [comment, marked as answer]
There is no procedure, since not supported.

**Reply by sfromis** [reply]
Opening up a Tasmota device to the Intenet, DDNS or not, is strongly not recommended

**Comment by rtek1000** [comment]
Since eWeLink uses the same hardware over the Internet, if Tasmota is not recommended to be used over the Internet, then Tasmota is so insecure that it totally relies on the security of the wifi network. And we know that the Wifi network is not completely secure.

> Wi-Fi encryption developed yet another chink in its armor this week. It’s now much easier to grab the hashed key.
> 
> So a hacker can capture a ton of WPA2 traffic, take it away, and decrypt it offline.

https://techbeacon.com/security/wpa2-hack-allows-wi-fi-password-crack-much-faster

**Comment by barbudor** [comment]
DDNS is NOT a standard, each DDNS provider may have it's own API, have it's own endpoint.
So there's no point in implementing a feature that would work with only a few vendors.

Your selected vendor should provide you with the HTTP API to call to set your DDNS.
There is no difficulties on Tasmota to use WebQuery to retrieve your external IP address from a service like `http://ifconfig.me/all.json` and then use another WebQuery to setup your DDNS.

You are claiming that eWelink is allowing access through the Internet which is totally wrong and absolutely not related to DDNS. 
All communications between the mobile eWelink/Tuya & al APP and the devices are going through the cloud infrastructure.

There is no difficulties to use a mobile APP with Tasmota using a free MQTT/TLS broker in the cloud. I'm using Flespi.io on a site where I didn't want to install a server and using Android app IoT MQTT Panel.
Which also allow me to integrate this remote site into my main home infrastructure (node-red/InfluxDB/grafana) seamlessly without having to care about openning ports nor needing DDNS.

**Reply by rtek1000** [reply]
barbudor says:
> You are claiming that eWelink is allowing access through the Internet which is totally wrong and absolutely not related to DDNS.
> All communications between the mobile eWelink/Tuya & al APP and the devices are going through the cloud infrastructure.

I found this:

> Network
> Cloud resources are typically delivered to users over the internet, so there is a need for third-party service providers to build and maintain the networking infrastructure that makes this possible. This infrastructure includes physical wiring, switches, load balancers and routers that help ensure cloud infrastructure is always available for customers when needed.

https://www.sumologic.com/glossary/cloud-infrastructure/

> What is eWeLink (4 years ago Updated)
> IoT has been trendy lately. With the development of smartphone and tablets, people expect all other devices that use electricity to connect to their home network and be at their command remotely. This accelerates the leap from the era of Mobile Web to Internet of Things.
> 
>   Ever more hardware companies release their own smart hardware that can connect to the internet. While still it is not feasible for many traditional manufacturers to build their Global IoT platform. To seize the chance of rapid IoT development,they opt to connect their devices to eWeLink’s platform. By connecting to eWeLink platform and APP, these products of different types from various brands are put under control and management in one app. Furthermore, eWeLink partners with multiple voice assistant producers and makes it easy to voice control these devices via Voice assistant such as Google assistant, Alex and more to come soon.

https://help.ewelink.cc/hc/en-us/articles/360005149451-What-is-eWeLink

**Comment by Jason2866** [comment]
You can not compare ewelink software with Tasmota. We dont know on which framework it is based.
Assuming that this software is secure is a hope. You can't check since it is closed source.
Tasmota, is NOT designed to be connected directly to the Internet. So yes it is insecure.
Tasmota is designed to control devices fully local. Just this, not more not less.
Btw. What ewelink is doing is out of any interest for us.

**Reply by rtek1000** [reply]
It would be nice to have a little security, as the wi-fi network can be observed outside the property boundaries, not all users who will be using Tasmota on a farm far from other neighbors can collect data from the network.

**Reply by Jason2866** [reply]
You can not secure a Arduino framework based ESP8266 device to be secure.
You have only the WiFi security. If this is not enough, do not use.

**Reply by rtek1000** [reply]
Jason2866 says:
> You can not secure a Arduino framework based ESP8266 device to be secure.

I found this:

> In this example we learn how to send Secured GET request to a secured https web page using NodeMCU or ESP8266? As we know all web pages are HTTP protocols, GET and POST are methods of communicating between web browser and the server. Also you look towards server side php Example coding.

https://circuits4you.com/2019/01/10/esp8266-nodemcu-https-secured-get-request/

Jason2866 says:
> If this is not enough, do not use.

I think not all people think like this: "Yes, don't use it. And fold your arms, don't improve anything, simple as that"

**Reply by rtek1000** [reply]
> Firmware
> Arduino core for ESP8266 WiFi chip V2.3.0 firmware adds HTTPS (TLS 1.0 and 1.1) support, making this a viable platform for secure IoT data streaming. See Security Discussion for more information.

https://github.com/gloveboxes/Arduino-ESP8266-Secure-Http-Azure-IoT-Hub-Client-V2/blob/master/README.md

**Reply by Jason2866** [reply]
Yes, based on a framework (nonOS SDK) which is closed source.
You can secure it like beeing a little pregnant.
Knowing it is based on insecure base, it would be a false signal.
A chain is only so strong as the weakest part. So we treat Tasmota by design as insecure.
The point is, we tried to explain why DDNS will not be considered to be implemented.
We see Tasmota (because of the hardware and the underlying Software) as not secure enough
to have this feature.

**Reply by sfromis** [reply]
Using https for planned communication does not mean that a device is safe to expose to the Internet allowing incoming connections. Outgoing https is the easy part, but does not make a server secure.
